InkyPhat/app/src/main/java/com/novoda/inkyphat/InkyPhatTriColourDisplay.java

```diff
@@ -0,0 +1,188 @@
+         pixelBuffer.setImage(x, y, imageConverter.convertText(text, color));
+    }
+
+    @Override
+    public void setPixel(int x, int y, int color) {
+        pixelBuffer.setPixel(x, y, colorConverter.convertARBG888Color(color));
+    }
+
+    @Override
+    public void setBorder(Palette color) {
+        switch (color) {
+            case BLACK:
+                border = BORDER_BLACK;
+                break;
+            case RED:
+                border = BORDER_RED;
+                break;
+            case WHITE:
+                border = BORDER_WHITE;
+                break;
+            default:
+                throw new IllegalStateException(color + " is unsupported as a border");
+        }
+    }
+
+    @Override
+    public void refresh() {
+        try {
+            turnDisplayOn();
+            update();
+            turnDisplayOff();
+        } catch (IOException e) {
+            throw new IllegalStateException("cannot init", e);
+        }
+    }
+
+    private void turnDisplayOn() throws IOException {
+        busyWait();
+
+        sendCommand(POWER_SETTING, new byte[]{0x07, 0x00, 0x0A, 0x00});
+        sendCommand(BOOSTER_SOFT_START, new byte[]{0x07, 0x07, 0x07});
+        sendCommand(POWER_ON);
+
+        busyWait();
+
+        sendCommand(PANEL_SETTING, new byte[]{(byte) 0b11001111});
+        sendCommand(VCOM_DATA_INTERVAL_SETTING, new byte[]{(byte) (0b00000111 | border)});
+
+        sendCommand(OSCILLATOR_CONTROL, new byte[]{0x29});
+        sendCommand(RESOLUTION_SETTING, new byte[]{0x68, 0x00, (byte) 0xD4});
+        sendCommand(VCOM_DC_SETTING, new byte[]{0x0A});
+
+        sendCommand(PARTIAL_EXIT);
+    }
+
+    private void update() throws IOException {
+        sendCommand(DATA_START_TRANSMISSION_1, pixelBuffer.getDisplayPixelsForColor(Palette.BLACK));
+        sendCommand(DATA_START_TRANSMISSION_2, pixelBuffer.getDisplayPixelsForColor(Palette.RED));
+        sendCommand(DISPLAY_REFRESH);
+    }
+
+    private void turnDisplayOff() throws IOException {
+        busyWait();
+
+        sendCommand(VCOM_DATA_INTERVAL_SETTING, new byte[]{0x00});
+        sendCommand(POWER_SETTING, new byte[]{0x02, 0x00, 0x00, 0x00});
+        sendCommand(POWER_OFF);
+    }
+
+    private void sendCommand(byte command) throws IOException {
+        chipCommandPin.setValue(SPI_COMMAND);
+        byte[] buffer = new byte[]{command};
+        spiBus.write(buffer, buffer.length);
+    }
+
+    private void sendCommand(byte command, byte[] data) throws IOException {
+        sendCommand(command);
+        sendData(data);
+    }
+
+    private void sendData(byte[] data) throws IOException {
+        chipCommandPin.setValue(SPI_DATA);
+        spiBus.write(data, data.length);
+    }
+
+    /**
+     * Wait for the e-paper driver to be ready to receive commands/data.
+     *
+     * @throws IOException error accessing GPIO
+     */
+    private void busyWait() throws IOException {
+        while (true) {
```

Comment:
or the code became out of sync and `busyWait` was called without resetting the pin its meant to be watching